Introduce `TypeMemberOrder` check

benhalasi opened the pull request:
#595

### Suggested commit message
```
Introduce `TypeMemberOrder` check (#636)
```

benhalasi committed:
Introduce `TypeMemberOrder` check

benhalasi commented:
Verifying that the whitespace is untouched may be a bit harder than I thought because [`TestMode.TEXT_MATCH` formats both input and output](https://github.com/google/error-prone/blob/31d420b282603d92ad105defed2f47c57ffd1691/test_helpers/src/main/java/com/google/errorprone/BugCheckerRefactoringTestHelper.java#L95)

rickie reviewed (commented):
There were some imports from `ASTHelpers` that we usually do not statically import :).

Added a commit! The setup looks nice, didn't get till the end, but already flushing my comments 😄.

rickie commented on error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/MemberOrdering.java:
```diff
@@ -0,0 +1,217 @@
+        .flatMap(List::stream);
+  }
+
+  // XXX: From this point the code is just a fancy copy of `SuggestFixes.replaceIncludingComments` -
```
We can indeed drop this `replaceIncludingComments` method! Will push :).

oxkitsune commented on error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/MemberOrdering.java:
```diff
@@ -0,0 +1,196 @@
+    for (Tree member : classTree.getMembers()) {
+      ImmutableList<ErrorProneToken> memberTokens =
+          ErrorProneTokens.getTokens(state.getSourceForNode(member), state.context);
+      if (memberTokens.isEmpty() || memberTokens.get(0).kind() == Tokens.TokenKind.EOF) {
+        continue;
+      }
+
+      @Var
+      ImmutableList<ErrorProneToken> maybeCommentedMemberTokens =
+          ImmutableList.copyOf(tokens.subList(0, memberTokens.size()));
+      while (!areTokenListsMatching(memberTokens, maybeCommentedMemberTokens)) {
+        tokens.remove(0);
+        maybeCommentedMemberTokens = ImmutableList.copyOf(tokens.subList(0, memberTokens.size()));
+      }
+
+      membersWithComments.add(
+          new MemberWithComments(
+              member, ImmutableList.copyOf(maybeCommentedMemberTokens.get(0).comments())));
+    }
+    return membersWithComments.build();
+  }
+
+  /**
+   * Checks whether two lists of error-prone tokens are 'equal' without considering their comments.
+   */
+  private static boolean areTokenListsMatching(
+      ImmutableList<ErrorProneToken> tokens, ImmutableList<ErrorProneToken> memberTokens) {
+    if (tokens.size() != memberTokens.size()) {
+      return false;
+    }
+    for (int i = 0; i < tokens.size() - 1 /* EOF */; i++) {
+      if (tokens.get(i).kind() != memberTokens.get(i).kind()
+          || tokens.get(i).hasName() != memberTokens.get(i).hasName()
+          || (tokens.get(i).hasName()
+              && !tokens.get(i).name().equals(memberTokens.get(i).name()))) {
+        return false;
+      }
+    }
+    return true;
+  }
```
```suggestion
  /** Returns the class' members with their comments. */
  private static ImmutableList<MemberWithComments> getMembersWithComments(
      ClassTree classTree, VisitorState state) {
    List<ErrorProneToken> tokens =
        state.getOffsetTokens(
            ASTHelpers.getStartPosition(classTree), state.getEndPosition(classTree));

    ImmutableList.Builder<MemberWithComments> membersWithComments = ImmutableList.builder();
    for (Tree member : classTree.getMembers()) {
      int memberStartPos = ASTHelpers.getStartPosition(member);
      // we start and the end position of the previous token
      // this gives us a range that includes any comments
      int startPos =
          tokens.stream()
              .map(ErrorProneToken::endPos)
              .filter(i -> i < memberStartPos)
              .reduce((first, second) -> second)
              .orElse(memberStartPos);

      ImmutableList<ErrorProneToken> memberTokens =
          ImmutableList.copyOf(state.getOffsetTokens(startPos, state.getEndPosition(member)));
      if (memberTokens.isEmpty() || memberTokens.get(0).kind() == Tokens.TokenKind.EOF) {
        continue;
      }

      membersWithComments.add(
          new MemberWithComments(member, ImmutableList.copyOf(memberTokens.get(0).comments())));
    }
    return membersWithComments.build();
  }
```
`ErrorProneTokens#getTokens(String, Context)` doesn't include the comments because it uses the start position of the provided member, e.g. the position of `void`. And while comments are not tokens, they do offset the position of tokens.

Using `VisitorState#getOffsetTokens(int, int)` we're able to explicitly choose the start/end positions of the token range. If we start the range at the end position of the previous token and end it at the end of the current member, we get the same tokens, but since the range now includes the offset for any comments these will be included in the resulting `ErrorProneToken`.

oxkitsune commented:
`ErrorProneTokens#getTokens(String, Context)` doesn't include comments as it relies on the start position of the provided member (e.g., the position of `void`). Comments don't count as tokens but do affect token positions.

To include comments, we can use `VisitorState#getOffsetTokens(int, int)` to obtain tokens in a specific range. By starting the range at the end position of the previous token and ending it at the end of the current member, we'll obtain the same tokens, now including comments in the resulting `ErrorProneToken`.

benhalasi commented:
Thanks oxkitsune for the `ErrorProneTokens#getTokens(String, Context)`!

benhalasi commented on error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/MemberOrdering.java:
```diff
@@ -0,0 +1,187 @@
+      return ImmutableList.of();
+    }
+
+    checkState(
+        state.getEndPosition(member) != -1,
+        "Member's end position is not available (-1).\n - member=[%s]\n - source=[%s]",
+        member,
+        state.getSourceForNode(member));
```
100% overkill, but it would be nice to not forget about this case or verify that it can never happen.

oxkitsune replied:
I ran into this exact issue while testing, to my knowledge this case can only happen for generated members (e.g. generated constructors)

benhalasi replied:
I was thinking that Lombok and tools like it could also behave like this.

benhalasi replied:
This makes two of us, I've applied changes assuming that `state.getEndPosition(member)` is only unavailable for members that are not in the source file.

benhalasi commented on error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/MemberOrdering.java:
```diff
@@ -0,0 +1,187 @@
+  /** A comparator that sorts variable and method (incl. constructors) in a standard order. */
+  private static final Comparator<Tree> MEMBER_SORTING =
+      comparing(
+          (Tree tree) -> {
```
Saw this being renamed to `tree` from `memberTree`, but in other placed we renamed `memberTree` to `classMemberTree`.

Personally I don't mind either. I originally went with member tree because afaik, only classes have members in Java, specifying `class` would be redundant. One thing to note, if we go with `classMember*`, we should rename this BugChecker to `ClassMemberOrdering`

oxkitsune replied:
> Saw this being renamed to tree from memberTree while in other places we renamed memberTree to classMemberTree. Do you think we can use the same strategy everywhere?

I think this is purely based on the type we're referring to, e.g. `Tree` -> `tree`.
And since we renamed `MemberTreeWithComments` to `ClassMemberTreeWithComments` the variable names should reflect that.

> I originally went with member tree because afaik, only classes have members in Java, specifying class would be redundant.

`ClassMemberTreeWithComments` is a name @rickie preferred. I think it does add some clarity. That being said, [JEP 445](https://openjdk.org/jeps/445) introduces unnamed classes which could be a concern for the `MemberTreeWithComments` name :joy:

benhalasi commented on error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/MemberOrdering.java:
```diff
@@ -0,0 +1,187 @@
+    Optional<Integer> previousMemberEndPos =
+        tokens.stream()
+            .map(ErrorProneToken::endPos)
+            .takeWhile(endPos -> endPos <= memberStartPos)
```
I did some investigation on the `<` or `<=` question. I don't think, it matters too much which way we go. It would matter if we could place a comment on a token that immediately follows another. This is impossible because if we comment a token, the token immediately follows its comment instead of its prior token.

```
// `int b` immidately follows the token `;`
int a;int b;
// `int b` doesn't follow immidately any error-prone token.
int a;/* comment */int b;
```
An interesting case would be a code like this:
```
...
/* comment */
;void method() { ... }
```
In this case, the comment belongs to the token `;` and not the method. Our current implementation will move the comment with the `void method() { ... }` instead of leaving it with the `;` token. This aligns with the implementation of `SuggestedFixes#replaceIncludingComments` and it provides a better UX than the alternative, but AFAIK it is technically incorrect.

I've added a test case for this, but again, this will be resolved by the simplest formatters.

oxkitsune replied:
Interesting case! 
I originally used `<=` over `<`, as I figured there was no reason to exclude the start position of the token, if for some reason a previous token has that as end position.

benhalasi commented on error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/MemberOrdering.java:
```diff
@@ -0,0 +1,187 @@
+    if (member.getKind() == Tree.Kind.METHOD
+        && ASTHelpers.isGeneratedConstructor((MethodTree) member)) {
+      return ImmutableList.of();
+    }
```
See thread above, I've replaced this condition with `state.getEndPosition(member) == -1` as the previous solution might have broke projects that use Lombok or tools like it.

oxkitsune replied:
Ah I forgot about Lombok :joy:. SGTM

oxkitsune commented:
Looks good!

However blocks (both static and instance) are also considered class members:
```java
class Foo {
	private static final int bar;
	private final int baz;


	// some static initializer block
	static {
		
		bar = 110 ^ 67;
	}

	// some instance initializer block
	{
		baz = (1395 ^ 1059) >> 3;
	}
}
```

Currently these are not sorted, I'm unsure what their order should be though. 
Something like this perhaps?
- static fields
- static *blocks*
- non-static fields
- non-static *blocks*

WDYT :smile:

benhalasi commented:
> However blocks (both static and instance) are also considered class members: ...

Good point, they're indeed non-handled. My reasoning to ignore this case was that using static blocks are very rare and specific. It would be hard to put these very specific edge cases into a rule, if even possible. Seems more practical to not sort them in this rule (or at all).

Afaik the general opinion on static blocks is that we should just not use them. Considering this, (imo) if we want to do anything with static blocks in error-prone, we should just flag them as errors.

Stephan202 reviewed (commented):
Rebased and added a commit. Didn't finish reviewing yet; timing of next round is TBD. Nice work! :rocket:

Stephan202 reviewed (commented):
Apparently back in October I didn't push my review comments :facepalm:. Doing so now; not sure what the state of this PR is; would need to get back to that some other time.

benhalasi commented:
Hehe, I'll check it out *soon* (this week 🤞) and will update it with the open questions!

benhalasi commented on error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/ClassMemberOrdering.java:
```diff
@@ -0,0 +1,181 @@
+  }
+
+  @AutoValue
+  abstract static class ClassMemberWithComments {
```
Considering the `Class -> Type` rename we're doing, wdyt of `TypeMember` or just `Member`?

Asking because `TypeMember` sounded weird to me initially. Probably I'm just not used to it.

Stephan202 replied:
Yeah, `TypeMember` can work! (Doesn't sound so weird to me :smile:.)

benhalasi commented:
Seems PiTest in CI consistently fails in ~1m w/
```
Failed to execute goal org.pitest:pitest-maven:1.15.3:mutationCoverage (default-cli) on project error-prone-contrib: Execution default-cli of goal org.pitest:pitest-maven:1.15.3:mutationCoverage failed: Coverage generation minion exited abnormally! (UNKNOWN_ERROR)
```
Didn't deep-dive in its logs nor tried executing it locally yet, just flagging!

benhalasi commented on error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/TypeMemberOrdering.java:
```diff
@@ -0,0 +1,236 @@
+      return ImmutableList.of();
+    }
+
+    // TODO: Move identifying "previous member end position" to an outer loop,
+    //  Loop once and identify for all members
+    // TODO: Check if this handles properly comments on the first member.
+    Optional<Integer> previousMemberEndPos =
+        state.getOffsetTokens(typeStart, typeEnd).stream()
+            .map(ErrorProneToken::endPos)
+            .takeWhile(endPos -> endPos < memberStart)
+            .reduce((earlierPos, laterPos) -> laterPos);
+
+    List<ErrorProneToken> typeMemberTokens =
+        state.getOffsetTokens(previousMemberEndPos.orElse(memberStart), memberEnd);
```
This TODO is for addressing the very not performant ways of this solution.

The challenge here to keep the comment and maybe other code too between members.

Options (I can think of atm):

---

- Somehow get all `ErrorProneTokens` (or source) of the class,
- divide them for each member, and
- reorder these token+member groups.

I believe this was described in this [comment](https://github.com/PicnicSupermarket/error-prone-support/pull/636#discussion_r1362763320).

With this I'm not sure how we can handle the whitespace properly, but I'm yet to fully discover this option.

---

- Go through the type's tokens once,
- gather all commented tokens, and put them in a `Map<Identifier, Comments>`
- use this map to get the comments of a member (`tree.getMembers()`)

I tried this quickly, using `ErrorProneToken#pos` and `ASTHelpers.getStartPosition(member)` as identifier.

This worked in general, but failed the `detached comment from method` test and the `replacementSuggestedFixDoesNotModifyWhitespace`.

---
---

Additional Note:

`VisitorState#getOffsetTokens` java doc says:
> This is moderately expensive (the source of the node has to be re-lexed), so it should only be used if a fix is already going to be emitted.

With this in mind, we should first determine if the members are ordered properly, and only then fetch the comments for them - if we opt to use this helper.

benhalasi reviewed (commented):
Open questions:
 - Specifically how to obtain the comments of a member / source between two members - here I have some ideas I'll try them in a timely manner :D
 - Whether regular identification and re-placement tests are sufficient to fully cover our logic - I believe this is highly dependent on how we resolve the comments.
- And some seemingly minor suggestions open to address.

Stephan202 commented:
> Seems PiTest in CI consistently fails in ~1m w/
> 
> ```
> Failed to execute goal org.pitest:pitest-maven:1.15.3:mutationCoverage (default-cli) on project error-prone-contrib: Execution default-cli of goal org.pitest:pitest-maven:1.15.3:mutationCoverage failed: Coverage generation minion exited abnormally! (UNKNOWN_ERROR)
> ```
> 
> Didn't deep-dive in its logs nor tried executing it locally yet, just flagging!

If you rebase this should be fixed, thanks to #976 :)

W.r.t. other questions: I'll have to get back to you, hopefully later this week. Tnx for reviving this PR!

Stephan202 commented:
Hey @benhalasi! I figured out why I hadn't pushed that October review until recently. I found a local branch with unpushed changes; apparently I just abandoned the review midway. I suspect that by now you've applied the relevant feedback, but just to be sure I pushed a commit with my local changes: 997099e957ecc2281b75b652d45fa8a3a0b9aaa1; perhaps there's something useful in it still (didn't check).

Will circle back to this PR later!

benhalasi commented:
Small progress update: I rewrote most of the checker's logic, (much inspired by https://github.com/PicnicSupermarket/error-prone-support/commit/997099e957ecc2281b75b652d45fa8a3a0b9aaa1).

The main changes:
- track unmovable members too so we can track source-code between members - including unmovable ones too.
- accommodate enums. I left `XXX: To accommodate enums` comments at their support's main pillars, and it's not even that horrifying!
- combine `BY_PREFERRED_TYPE_MEMBER_ORDER` Comparator with the logic that determines which members are movable.
- split the tests between classes, interfaces, and enums.

I'm yet to sleep one on it and have a final round of changes mark parts where I have minor questions, but ofc any input is appreciated!

rickie commented on error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/TypeMemberOrder.java:
```diff
@@ -0,0 +1,265 @@
+    return describeMatch(tree, sortTypeMembers(bodyStartPos, members, state));
+  }
+
+  private static boolean isGeneratedConstructor(Tree tree) {
```
We can maybe move this to `MoreAstHelpers` 🤔 , although it's used only once right now..

WDYT of this implementation:
```
  public static boolean isGeneratedConstructor(Tree tree) {
    return tree instanceof MethodTree methodTree && ASTHelpers.isGeneratedConstructor(methodTree);
  }
```

rickie replied:
It's inversed, it returns true if `Tree` is  not of type `Kind.METHOD`, will push a commit to fix :).

rickie commented:
/integration-test

rickie commented on error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/TypeMemberOrder.java:
```diff
@@ -0,0 +1,265 @@
+ *     href="https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/checks/coding/DeclarationOrderCheck.html">Checkstyle's
+ *     {@code DeclarationOrderCheck}</a>
+ */
+// XXX: Reference
```
I think we can drop this :).

rickie commented on error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/TypeMemberOrder.java:
```diff
@@ -0,0 +1,265 @@
+// https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/checks/coding/DeclarationOrderCheck.html
+@AutoService(BugChecker.class)
+@BugPattern(
+    summary = "Type members should be ordered in a standard way",
```
```suggestion
    summary = "Type members should be defined in a canonical order",
```
What about this?

rickie commented on error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/TypeMemberOrder.java:
```diff
@@ -0,0 +1,265 @@
+      return Description.NO_MATCH;
+    }
+
+    // All members that can be moved or may lay between movable ones.
```
If we leave comments in the class, we usually have the `/* */` style.

rickie commented on error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/TypeMemberOrderInterfaceTest.java:
```diff
@@ -0,0 +1,198 @@
+  }
+
+  @Test
+  void replacementSuggestedFixConsidersDanglingComments() {
```
I think we can simplify the name of the tests :).

rickie reviewed (commented):
We discussed this earlier, but I think it would be nice to introduce this check in the `error-prone-experimental` module. From there we continue validation and move it to `error-prone-contrib` after that.

Man, this check is really cool but takes some time to review. Did another pass and had some style suggestions/tweaks. Please take a look and feel free to tweak :).

rickie commented on error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/TypeMemberOrder.java:
```diff
@@ -0,0 +1,260 @@
+   */
+  private static int getBodyStartPos(ClassTree tree, VisitorState state) {
+    CharSequence sourceCode = state.getSourceCode();
+    int typeStart = ASTHelpers.getStartPosition(tree);
```
Usually we call this `startPosition`, I think we can do the same here 🤔, not sure if we should make it specific to add "type"? Or am I missing a reason to do so?

rickie commented:
/integration-test

rickie commented:
/integration-test

rickie reviewed (commented):
Sorry I didn't put comments everywhere. The main idea behind the changes is that we now do only one "scan" and collect only the members that we can actually change. After that we check is the list sorted, and if not we do the same as before. But now, we know for sure that we only have items that we can and want to switch around :). Makes it a bit easier as we don't need to check for empty/present optionals and have all the data. 

So there is still a specific case not working and I didn't finish reviewing all the tests yet, but man this will be an epic check.

Oh and we should still move this to experimental :).

Credits to @Stephan202 as we paired on this.

rickie commented:
/integration-test

benhalasi commented:
/integration-test

rickie commented:
It's not working as this check is in the experimental module 😬. So we will need to tweak something, or to run it locally 😄 @benhalasi.

benhalasi committed:
Introduce `Whitespace` check

rickie committed:
Apply EPS suggestions

rickie committed:
Pair review

rickie committed:
Pair review 2

rickie committed:
Pair review 3

benhalasi committed:
Tweaks

rickie committed:
Format

rickie committed:
Fix the dependencies

rickie committed:
Small tweaks

rickie commented:
/integration-test

rickie committed:
Small tweaks
